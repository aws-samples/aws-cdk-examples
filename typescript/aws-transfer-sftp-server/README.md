# AWS CDK Example for deploying AWS Transfer SFTP Server

<!--BEGIN STABILITY BANNER-->
---

![Stability: Stable](https://img.shields.io/badge/stability-Stable-success.svg?style=for-the-badge)

> **This is a stable example. It should successfully build out of the box**
>
> This example is built on Construct Libraries marked "Stable" and does not have any infrastructure prerequisites to build.
---
<!--END STABILITY BANNER-->

## AWS Transfer Family

AWS Transfer Family is a secure transfer service that enables you to transfer files into and out of AWS storage services. AWS Transfer Family supports [Secure Shell (SSH) File Transfer Protocol (SFTP): version 3](https://docs.aws.amazon.com/transfer/latest/userguide/what-is-aws-transfer-family.html).

File transfer protocols are used in data exchange workflows across different industries such as financial services, healthcare, advertising, and retail, among others. Transfer Family simplifies the migration of file transfer workflows to AWS.

With Transfer Family, you get access to a file transfer protocol-enabled server in AWS without the need to run any server infrastructure. You can use this service to migrate your file transfer-based workflows to AWS while maintaining your end users' clients and configurations as is.

## Solution overview

The example will deploy an [SFTP server](https://docs.aws.amazon.com/transfer/latest/userguide/create-server-sftp.html) with a single user and configures a public key for the [user](https://docs.aws.amazon.com/transfer/latest/userguide/service-managed-users.html).

A security group can optionally be configured to allowlist only specific IPs.

S3 bucket is created for storing the data. User is scoped to only have access to a specific prefix in the bucket.

![Architecture diagram of the solution](./img/aws-transfer-sftp-server.png)

## Build and test

First, follow the [CDK getting started guide](https://docs.aws.amazon.com/cdk/v2/guide/getting_started.html).

To build this app, you need to be in this example's root folder. Then run the following:

```sh
npm install -g aws-cdk
npm install
```

This will install the latest version of CDK and other dependencies.

Then verify that the snapshot test succeeds:

```sh
npm run test
```

## Deploy

First, review the `main.ts` to see if you want to configure account, region or SFTP user name.
Proceed to run CDK deploy with [context variable](https://docs.aws.amazon.com/cdk/v2/guide/context.html) `userPublicKeys` and optionally also `allowedIps`.

```sh
cdk deploy --context "userPublicKeys=ecdsa-sha2-nistp256 AAAAE2VjZHNhLXNoYTItbmlzdHAyNTYAAAAIbmlzdHAyNTYAAABBBCJlxu5q1M3icgvrvNvCyE4gavDWaB8L7ZyGjnpsp/7GZhczaqY49KmZnZrbsKfoKtKu5bkNN8BXcjrAAwwv0Hk="
```

When running the deployment in CI/CD environment, you likely want to fetch the context from your own parameter or secrets store.

The command will deploy or update the stack on the active AWS account.
`cdk destroy` can be used to remove the deployed resources after they are no longer needed.

## Connect

Test the server using SFTP client:

```sh
$ sftp -i ~/.ssh/id_ecdsa sftp-user-dev@s-server-id.server.transfer.eu-north-1.amazonaws.com
sftp> ls
sftp> put example_file
sftp> ls
example_file
sftp>
```
