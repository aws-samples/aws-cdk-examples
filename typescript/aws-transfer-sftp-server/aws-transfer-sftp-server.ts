import { Stack, StackProps, RemovalPolicy, CfnOutput } from 'aws-cdk-lib';
import {
  aws_s3 as s3,
  aws_ec2 as ec2,
  aws_iam as iam,
  aws_logs as logs,
  aws_cloudwatch as cw,
  aws_transfer as transfer,
} from 'aws-cdk-lib';
import { Construct } from 'constructs';

/** Properties required to setup the SFTP server. */
export interface SftpServerStackProps extends StackProps {
  /** User name for the default SFTP user.
   * @example 'sftp-user'
   */
  userName: string;

  /** The public key of the SFTP user. */
  userPublicKeys: string[];

  /** IP addresses that are allowed to connect to the SFTP server.
   * @default All IPV4 is allowed.
   */
  allowedIps?: string[];

  /** The S3 bucket to be configured for the SFTP server. */
  dataBucket: s3.IBucket;
}

/** Stack for initializing a fully working SFTP server. */
export class SftpServerStack extends Stack {

  /** CloudWatch alarm that is triggered if there are too many errors in the logs. */
  errorAlarm: cw.Alarm;

  constructor(scope: Construct, id: string, props: SftpServerStackProps) {
    super(scope, id, props);

    const vpc = new ec2.Vpc(this, 'VPC', {
      maxAzs: 2,
      natGateways: 0,
    });

    // Create the required IAM role which allows the SFTP server
    // to log to CloudWatch.
    const cloudWatchLoggingRole = new iam.Role(this, 'CloudWatchLoggingRole', {
      assumedBy: new iam.ServicePrincipal('transfer.amazonaws.com'),
      description: 'IAM role used by AWS Transfer for logging',
      inlinePolicies: {
        loggingRole: new iam.PolicyDocument({
          statements: [new iam.PolicyStatement({
            actions: [
              'logs:CreateLogGroup',
              'logs:CreateLogStream',
              'logs:DescribeLogStreams',
              'logs:PutLogEvents',
            ],
            resources: [`arn:aws:logs:${this.region}:${this.account}:log-group:/aws/transfer/*`],
            effect: iam.Effect.ALLOW,
          })],
        }),
      },
    });

    // Security group for restricting incoming traffic to specific IP addresses
    const sg = new ec2.SecurityGroup(this, 'SftpServerSG', {
      vpc,
      allowAllOutbound: false,
      securityGroupName: 'SFTPServerSG',
      description: 'Security group for SFTP server',
    });

    // In production it's good to allow only specific IP addresses
    if (props.allowedIps) {
      props.allowedIps.forEach((ip) => {
        sg.addIngressRule(ec2.Peer.ipv4(ip), ec2.Port.tcp(22), 'Allow SSH inbound');
      });
    } else {
      sg.addIngressRule(ec2.Peer.anyIpv4(), ec2.Port.tcp(22), 'Allow SSH inbound');
    }

    // Create as many Elastic IP addresses as we have availability zones
    const eips = vpc.publicSubnets.map((_, index) => new ec2.CfnEIP(this, `SftpEIP${index + 1}`, {
      domain: 'vpc',
    }));

    const server = new transfer.CfnServer(this, 'SFTPServer', {
      endpointDetails: {
        securityGroupIds: [sg.securityGroupId],
        vpcId: vpc.vpcId,
        subnetIds: vpc.publicSubnets.map((subnet) => subnet.subnetId),
        addressAllocationIds: eips.map((eip) => eip.attrAllocationId),
      },
      identityProviderType: 'SERVICE_MANAGED',
      endpointType: 'VPC',
      loggingRole: cloudWatchLoggingRole.roleArn,
      protocols: ['SFTP'],
      domain: 'S3',
    });

    // Output Server Endpoint access where clients can connect
    new CfnOutput(this, 'SFTPServerEndpoint', {
      description: 'Server Endpoint',
      value: `${server.attrServerId}.server.transfer.${this.region}.amazonaws.com`,
    });

    // Allow SFTP user to write the S3 bucket
    const sftpAccessPolicy = new iam.ManagedPolicy(this, 'SftpAccessPolicy', {
      managedPolicyName: 'SftpAccessPolicy',
      description: 'SFTP access policy',
    });
    props.dataBucket.grantReadWrite(sftpAccessPolicy);

    const sftpUserAccessRole = new iam.Role(this, 'SftpAccessRole', {
      assumedBy: new iam.ServicePrincipal('transfer.amazonaws.com'),
      roleName: 'SftpAccessRole',
      managedPolicies: [
        sftpAccessPolicy,
      ],
    });

    const logGroup = new logs.LogGroup(this, 'SftpLogGroup', {
      logGroupName: `/aws/transfer/${server.attrServerId}`,
      removalPolicy: RemovalPolicy.DESTROY,
      retention: logs.RetentionDays.ONE_MONTH,
    });

    // Configure user which has access to the S3 bucket
    // https://docs.aws.amazon.com/transfer/latest/userguide/service-managed-users.html
    new transfer.CfnUser(this, 'SFTPUser', {
      serverId: server.attrServerId,
      homeDirectory: `/${props.dataBucket.bucketName}/incoming-data`,
      role: sftpUserAccessRole.roleArn,
      userName: props.userName,
      sshPublicKeys: props.userPublicKeys,
    });

    // Metric filter for recognizing two types of errors in the SFTP logs
    const metricFilter = new logs.MetricFilter(this, 'MetricFilter', {
      logGroup,
      metricNamespace: 'SftpServer',
      metricName: 'ErrorLog',
      filterPattern: logs.FilterPattern.anyTerm('ERRORS AUTH_FAILURE', 'ERROR Message'),
      metricValue: '1',
      unit: cw.Unit.COUNT,
    });

    // Alarm if there are too many errors
    this.errorAlarm = new cw.Alarm(this, 'AlarmMetricFilter', {
      alarmDescription: 'Alarm if there are too many errors in the logs',
      metric: metricFilter.metric(),
      threshold: 1,
      evaluationPeriods: 5,
      datapointsToAlarm: 1,
    });

    // TODO Add alarm action to notify administrators or perform other actions
  }
}
