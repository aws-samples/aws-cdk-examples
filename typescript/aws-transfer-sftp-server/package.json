{
  "name": "aws-transfer-sftp-server",
  "version": "0.1.0",
  "bin": {
    "aws-transfer-sftp-server": "npx ts-node --prefer-ts-exts bin/main.ts"
  },
  "description": "AWS CDK Example for deploying AWS Transfer SFTP Server",
  "private": true,
  "author": {
    "name": "Amazon Web Services",
    "url": "https://aws.amazon.com",
    "organization": true
  },
  "license": "Apache-2.0",
  "scripts": {
    "build": "tsc",
    "watch": "tsc -w",
    "test": "jest",
    "cdk": "cdk"
  },
  "devDependencies": {
    "@types/jest": "^29.5.2",
    "@types/node": "18.15.11",
    "aws-cdk": "*",
    "jest": "^29.5.0",
    "ts-jest": "^29.1.0",
    "ts-node": "^10.9.1",
    "typescript": "~4.6.0"
  },
  "dependencies": {
    "aws-cdk-lib": "^2.0.0",
    "constructs": "^10.0.0"
  }
}
