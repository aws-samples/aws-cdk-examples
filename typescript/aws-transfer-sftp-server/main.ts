#!/usr/bin/env node
import * as cdk from 'aws-cdk-lib';
import {
  aws_s3 as s3,
} from 'aws-cdk-lib';
import { SftpServerStack, SftpServerStackProps } from './aws-transfer-sftp-server';

// Follow the setup process at https://docs.aws.amazon.com/cdk/v2/guide/environments.html
const props: cdk.StackProps = {
  env: {
    account: process.env.CDK_DEFAULT_ACCOUNT,
    region: process.env.CDK_DEFAULT_REGION,
  }
};

const app = new cdk.App();

// Use CDK Context to get configuration variables
// https://docs.aws.amazon.com/cdk/v2/guide/context.html
const publicKeysContext = app.node.tryGetContext('userPublicKeys') as string | undefined;
if (!publicKeysContext) {
  console.error('Parameter userPublicKeys missing');
}
const allowedIps = app.node.tryGetContext('allowedIps') as string | undefined;

const bucketStack = new cdk.Stack(app, 'IncomingDataStack-dev', props);
const bucket = new s3.Bucket(bucketStack, 'IncomingDataBucket', {
  blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
  bucketName: `sftp-server-data-bucket-${props.env?.account}-${props.env?.region}`,
  encryption: s3.BucketEncryption.KMS_MANAGED,
  enforceSSL: true,
  // Do not use for production!
  removalPolicy: cdk.RemovalPolicy.DESTROY,
});

const userPublicKeys = publicKeysContext ? publicKeysContext.split(',') : [];

// Different properties for dev/test environment
const sftpPropsDev: SftpServerStackProps = {
  userName: 'sftp-user-dev',
  userPublicKeys,
  dataBucket: bucket,
  ...props,
};

// In production we want to limit allowed IPs
const sftpProps: SftpServerStackProps = {
  userName: 'sftp-user',
  userPublicKeys,
  allowedIps: allowedIps ? allowedIps.split(',') : undefined,
  dataBucket: bucket,
  ...props,
};

new SftpServerStack(app, 'SftpServerStack-dev', sftpPropsDev);

new SftpServerStack(app, 'SftpServerStack-prod', sftpProps);
