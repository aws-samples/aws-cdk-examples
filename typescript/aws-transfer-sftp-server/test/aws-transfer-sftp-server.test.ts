import * as cdk from 'aws-cdk-lib';
import {
  aws_s3 as s3,
} from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { SftpServerStack } from '../aws-transfer-sftp-server';

test('SftpServerStack has required resources', () => {
  const app = new cdk.App();
  // WHEN
  const bucketStack = new cdk.Stack(app, 'IncomingDataStack-dev', { });
  const bucket = new s3.Bucket(bucketStack, 'IncomingDataBucket', {
    blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
    bucketName: `sftp-server-data-bucket-example`,
    encryption: s3.BucketEncryption.KMS_MANAGED,
    enforceSSL: true,
    // Do not use for production!
    removalPolicy: cdk.RemovalPolicy.DESTROY,
  });
  const stack = new SftpServerStack(app, 'SftpServerStack', {
    userName: 'sftp-user-dev',
    userPublicKeys: ['ecdsa-sha2-nistp256 AAAAE2VjZHNhLXNoYTItbmlzdHAyNTYAAAAIbmlzdHAyNTYAAABBBCJlxu5q1M3icgvrvNvCyE4gavDWaB8L7ZyGjnpsp/7GZhczaqY49KmZnZrbsKfoKtKu5bkNN8BXcjrAAwwv0Hk='],
    dataBucket: bucket,
  });

  expect(Template.fromStack(stack)).toMatchSnapshot();
});
