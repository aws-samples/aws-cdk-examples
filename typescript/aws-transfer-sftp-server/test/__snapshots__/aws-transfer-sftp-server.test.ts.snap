// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SftpServerStack has required resources 1`] = `
{
  "Outputs": {
    "SFTPServerEndpoint": {
      "Description": "Server Endpoint",
      "Value": {
        "Fn::Join": [
          "",
          [
            {
              "Fn::GetAtt": [
                "SFTPServer",
                "ServerId",
              ],
            },
            ".server.transfer.",
            {
              "Ref": "AWS::Region",
            },
            ".amazonaws.com",
          ],
        ],
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "AlarmMetricFilter9DBA3966": {
      "Properties": {
        "AlarmDescription": "Alarm if there are too many errors in the logs",
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "DatapointsToAlarm": 1,
        "EvaluationPeriods": 5,
        "MetricName": "ErrorLog",
        "Namespace": "SftpServer",
        "Period": 300,
        "Statistic": "Average",
        "Threshold": 1,
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "CloudWatchLoggingRole8E23D1D4": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "transfer.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "IAM role used by AWS Transfer for logging",
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": [
                    "logs:CreateLogGroup",
                    "logs:CreateLogStream",
                    "logs:DescribeLogStreams",
                    "logs:PutLogEvents",
                  ],
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:aws:logs:",
                        {
                          "Ref": "AWS::Region",
                        },
                        ":",
                        {
                          "Ref": "AWS::AccountId",
                        },
                        ":log-group:/aws/transfer/*",
                      ],
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "loggingRole",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "MetricFilter1B93B6E5": {
      "Properties": {
        "FilterPattern": "?"ERRORS AUTH_FAILURE" ?"ERROR Message"",
        "LogGroupName": {
          "Ref": "SftpLogGroupC79A244B",
        },
        "MetricTransformations": [
          {
            "MetricName": "ErrorLog",
            "MetricNamespace": "SftpServer",
            "MetricValue": "1",
            "Unit": "Count",
          },
        ],
      },
      "Type": "AWS::Logs::MetricFilter",
    },
    "SFTPServer": {
      "Properties": {
        "Domain": "S3",
        "EndpointDetails": {
          "AddressAllocationIds": [
            {
              "Fn::GetAtt": [
                "SftpEIP1",
                "AllocationId",
              ],
            },
            {
              "Fn::GetAtt": [
                "SftpEIP2",
                "AllocationId",
              ],
            },
          ],
          "SecurityGroupIds": [
            {
              "Fn::GetAtt": [
                "SftpServerSG79C38856",
                "GroupId",
              ],
            },
          ],
          "SubnetIds": [
            {
              "Ref": "VPCPublicSubnet1SubnetB4246D30",
            },
            {
              "Ref": "VPCPublicSubnet2Subnet74179F39",
            },
          ],
          "VpcId": {
            "Ref": "VPCB9E5F0B4",
          },
        },
        "EndpointType": "VPC",
        "IdentityProviderType": "SERVICE_MANAGED",
        "LoggingRole": {
          "Fn::GetAtt": [
            "CloudWatchLoggingRole8E23D1D4",
            "Arn",
          ],
        },
        "Protocols": [
          "SFTP",
        ],
      },
      "Type": "AWS::Transfer::Server",
    },
    "SFTPUser": {
      "Properties": {
        "HomeDirectory": {
          "Fn::Join": [
            "",
            [
              "/",
              {
                "Fn::ImportValue": "IncomingDataStack-dev:ExportsOutputRefIncomingDataBucket3554D8354A965D28",
              },
              "/incoming-data",
            ],
          ],
        },
        "Role": {
          "Fn::GetAtt": [
            "SftpAccessRoleBDBB5CE1",
            "Arn",
          ],
        },
        "ServerId": {
          "Fn::GetAtt": [
            "SFTPServer",
            "ServerId",
          ],
        },
        "SshPublicKeys": [
          "ecdsa-sha2-nistp256 AAAAE2VjZHNhLXNoYTItbmlzdHAyNTYAAAAIbmlzdHAyNTYAAABBBCJlxu5q1M3icgvrvNvCyE4gavDWaB8L7ZyGjnpsp/7GZhczaqY49KmZnZrbsKfoKtKu5bkNN8BXcjrAAwwv0Hk=",
        ],
        "UserName": "sftp-user-dev",
      },
      "Type": "AWS::Transfer::User",
    },
    "SftpAccessPolicy6C559C82": {
      "Properties": {
        "Description": "SFTP access policy",
        "ManagedPolicyName": "SftpAccessPolicy",
        "Path": "/",
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::ImportValue": "IncomingDataStack-dev:ExportsOutputFnGetAttIncomingDataBucket3554D835Arn7B042284",
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::ImportValue": "IncomingDataStack-dev:ExportsOutputFnGetAttIncomingDataBucket3554D835Arn7B042284",
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::ManagedPolicy",
    },
    "SftpAccessRoleBDBB5CE1": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "transfer.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Ref": "SftpAccessPolicy6C559C82",
          },
        ],
        "RoleName": "SftpAccessRole",
      },
      "Type": "AWS::IAM::Role",
    },
    "SftpEIP1": {
      "Properties": {
        "Domain": "vpc",
      },
      "Type": "AWS::EC2::EIP",
    },
    "SftpEIP2": {
      "Properties": {
        "Domain": "vpc",
      },
      "Type": "AWS::EC2::EIP",
    },
    "SftpLogGroupC79A244B": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/transfer/",
              {
                "Fn::GetAtt": [
                  "SFTPServer",
                  "ServerId",
                ],
              },
            ],
          ],
        },
        "RetentionInDays": 30,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Delete",
    },
    "SftpServerSG79C38856": {
      "Properties": {
        "GroupDescription": "Security group for SFTP server",
        "GroupName": "SFTPServerSG",
        "SecurityGroupEgress": [
          {
            "CidrIp": "255.255.255.255/32",
            "Description": "Disallow all traffic",
            "FromPort": 252,
            "IpProtocol": "icmp",
            "ToPort": 86,
          },
        ],
        "SecurityGroupIngress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow SSH inbound",
            "FromPort": 22,
            "IpProtocol": "tcp",
            "ToPort": 22,
          },
        ],
        "VpcId": {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "VPCB9E5F0B4": {
      "Properties": {
        "CidrBlock": "10.0.0.0/16",
        "EnableDnsHostnames": true,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": [
          {
            "Key": "Name",
            "Value": "SftpServerStack/VPC",
          },
        ],
      },
      "Type": "AWS::EC2::VPC",
    },
    "VPCIGWB7E252D3": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "SftpServerStack/VPC",
          },
        ],
      },
      "Type": "AWS::EC2::InternetGateway",
    },
    "VPCIsolatedSubnet1RouteTableAssociationA2D18F7C": {
      "Properties": {
        "RouteTableId": {
          "Ref": "VPCIsolatedSubnet1RouteTableEB156210",
        },
        "SubnetId": {
          "Ref": "VPCIsolatedSubnet1SubnetEBD00FC6",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VPCIsolatedSubnet1RouteTableEB156210": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "SftpServerStack/VPC/IsolatedSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VPCIsolatedSubnet1SubnetEBD00FC6": {
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            0,
            {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.128.0/18",
        "MapPublicIpOnLaunch": false,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Isolated",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Isolated",
          },
          {
            "Key": "Name",
            "Value": "SftpServerStack/VPC/IsolatedSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "VPCIsolatedSubnet2RouteTable9B4F78DC": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "SftpServerStack/VPC/IsolatedSubnet2",
          },
        ],
        "VpcId": {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VPCIsolatedSubnet2RouteTableAssociation7BF8E0EB": {
      "Properties": {
        "RouteTableId": {
          "Ref": "VPCIsolatedSubnet2RouteTable9B4F78DC",
        },
        "SubnetId": {
          "Ref": "VPCIsolatedSubnet2Subnet4B1C8CAA",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VPCIsolatedSubnet2Subnet4B1C8CAA": {
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            1,
            {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.192.0/18",
        "MapPublicIpOnLaunch": false,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Isolated",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Isolated",
          },
          {
            "Key": "Name",
            "Value": "SftpServerStack/VPC/IsolatedSubnet2",
          },
        ],
        "VpcId": {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "VPCPublicSubnet1DefaultRoute91CEF279": {
      "DependsOn": [
        "VPCVPCGW99B986DC",
      ],
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": {
          "Ref": "VPCIGWB7E252D3",
        },
        "RouteTableId": {
          "Ref": "VPCPublicSubnet1RouteTableFEE4B781",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "VPCPublicSubnet1RouteTableAssociation0B0896DC": {
      "Properties": {
        "RouteTableId": {
          "Ref": "VPCPublicSubnet1RouteTableFEE4B781",
        },
        "SubnetId": {
          "Ref": "VPCPublicSubnet1SubnetB4246D30",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VPCPublicSubnet1RouteTableFEE4B781": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "SftpServerStack/VPC/PublicSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VPCPublicSubnet1SubnetB4246D30": {
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            0,
            {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.0.0/18",
        "MapPublicIpOnLaunch": true,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          {
            "Key": "Name",
            "Value": "SftpServerStack/VPC/PublicSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "VPCPublicSubnet2DefaultRouteB7481BBA": {
      "DependsOn": [
        "VPCVPCGW99B986DC",
      ],
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": {
          "Ref": "VPCIGWB7E252D3",
        },
        "RouteTableId": {
          "Ref": "VPCPublicSubnet2RouteTable6F1A15F1",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "VPCPublicSubnet2RouteTable6F1A15F1": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "SftpServerStack/VPC/PublicSubnet2",
          },
        ],
        "VpcId": {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VPCPublicSubnet2RouteTableAssociation5A808732": {
      "Properties": {
        "RouteTableId": {
          "Ref": "VPCPublicSubnet2RouteTable6F1A15F1",
        },
        "SubnetId": {
          "Ref": "VPCPublicSubnet2Subnet74179F39",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VPCPublicSubnet2Subnet74179F39": {
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            1,
            {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.64.0/18",
        "MapPublicIpOnLaunch": true,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          {
            "Key": "Name",
            "Value": "SftpServerStack/VPC/PublicSubnet2",
          },
        ],
        "VpcId": {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "VPCVPCGW99B986DC": {
      "Properties": {
        "InternetGatewayId": {
          "Ref": "VPCIGWB7E252D3",
        },
        "VpcId": {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::VPCGatewayAttachment",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
